src/utils/Cookie.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of RakLib.
+ * Copyright (C) 2014-2022 PocketMine Team <https://github.com/pmmp/RakLib>
+ *
+ * RakLib is not affiliated with Jenkins Software LLC nor RakNet.
+ *
+ * RakLib is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ */
+
+declare(strict_types=1);
+
+namespace raklib\utils;
+
+use raklib\utils\InternetAddress;
+use pocketmine\utils\Binary;
+use function random_int;
+use function crc32;
+
+final class Cookie{
+
+	/**
+	 * @var array<string, int> $cookies
+	 */
+	private array $cookies = [];
+
+	public function get(InternetAddress $address) : ?int{
+		if (isset($this->cookies[$address->toString()])) {
+			return $this->cookies[$address->toString()];
+		}
+		return null;
+	}
+
+	public function check(InternetAddress $address, int $cookie) : bool{
+		$addressStr = $address->toString();
+
+		if (isset($this->cookies[$addressStr])) {
+			// If it checks the Cookie, it means that it is in the OpenConnectionRequest2 phase, and we can delete it from memory.
+			if ($this->cookies[$addressStr] === $cookie) {
+				unset($this->cookies[$addressStr]);
+				return true;
+			}
+		} // Is there any chance that this is something else?
+		return false;
+	}
+
+	public function add(InternetAddress $address) : void{
+		if (!isset($this->cookies[$address->toString()])) {
+			$this->cookies[$address->toString()] = $this->generate($address);
+		}
+	}
+
+	private function generate(InternetAddress $address) : int{
+		return crc32(Binary::writeLInt(random_int(0, 0xffffffff)) . Binary::writeLShort($address->getPort()) . $address->getIp());
```

Comment:
If you already have a random input into `crc32`, is there still any motivation to checksum it together with the client address?

Reply:
I mentioned this higher up, the whole thing makes no sense. `random_int()` by itself would be sufficient.